src/coreclr/scripts/genUserEvents.py

```diff
@@ -0,0 +1,407 @@
+                WriteEventImpl.append("}\n\n")
+            elif (runtimeFlavor.mono or runtimeFlavor.nativeaot):
+                WriteEventImpl.append("\n    return ERROR_SUCCESS;\n}\n\n")
+    return ''.join(WriteEventImpl)
+
+def generateWriteEventBody(template, providerPrettyName, eventName, runtimeFlavor, eventLevel, eventKeywordsMask):
+    #each of the if/else in this function is a different type of template, or a template containing specific args
+    fnSig = template.signature
+    pack_list = []
+    #need the providerNode to get the first arg for TraceLoggingWrite
+    pack_list.append("    TraceLoggingWriteActivity(%s, \"%s\", ActivityId, RelatedActivityId, TraceLoggingLevel(%s), TraceLoggingKeyword(%s)" % (providerPrettyName, eventName, getUserEventLogLevelMapping(runtimeFlavor)[eventLevel], eventKeywordsMask))
+
+    for paramName in fnSig.paramlist:
+        parameter = fnSig.getParam(paramName)
+        if paramName in template.structs:
+            size = "(int)%s_ElementSize * (int)%s" % (
+                paramName, parameter.prop)
+            pack_list.append("    TraceLoggingBinary(%s, %s)" % (parameter, size))
+        elif paramName in template.arrays:
+            size = "sizeof(%s) * (int)%s" % (
+                getArrayDataTypeMapping(runtimeFlavor)[parameter.winType],
+                parameter.prop)
+        elif parameter.winType == "win:GUID":
+            pack_list.append("    %s((uint8_t *)%s)" % (getUserEventDataTypeMapping(runtimeFlavor)[parameter.winType], parameter.name))
+        else:
+            pack_list.append("    %s(%s)" % (getUserEventDataTypeMapping(runtimeFlavor)[parameter.winType], parameter.name))
+    code = ",\n".join(pack_list)  + "\n    );\n    return ERROR_SUCCESS;\n" #+ "\n\n"
+    return code
+
+
+
+keywordMap = {}
+
+def generateEventKeywords(eventKeywords):
+    mask = 0
+    # split keywords if there are multiple
+    allKeywords = eventKeywords.split()
+
+    for singleKeyword in allKeywords:
+        mask = mask | keywordMap[singleKeyword]
+
+    return mask
+
+def getCoreCLRUserEventHelperFileImplPrefix():
+    return """
+#include "common.h"
+#include <stdlib.h>
+#include <string.h>
+
+#ifndef TARGET_UNIX
+#include <windef.h>
+#include <crtdbg.h>
+#else
+#include "pal.h"
+#endif //TARGET_UNIX
+
+"""
+
+def generateUserEventHelperFile(etwmanifest, userevent_directory, target_cpp, runtimeFlavor, extern, dryRun):
+    usereventhelpersPath = os.path.join(userevent_directory, "usereventshelpers" + (".cpp" if target_cpp else ".c"))
+    if dryRun:
+        print(usereventhelpersPath)
+    else:
+        with open_for_update(usereventhelpersPath) as helper:
+            helper.write(stdprolog_cpp)
+            helper.write("""
+#include <user_events.h>
+#include <common.h>
+#include <configuration.h>
+
+bool s_userEventsEnabled = false;
+
+void InitUserEvents()
+{
+    // TODO: What to call this runtimeconfig.json switch? S.D.T implies it supports EventSource
```

Comment:
I think we could just document this is an experimental feature and currently it does not support EventSource. Later we'd add EventSource support and leave the naming unchanged.